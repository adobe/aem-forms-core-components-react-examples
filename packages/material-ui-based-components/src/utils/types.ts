import { FieldJson, FieldsetJson, State } from '@aemforms/af-core';
import { Handlers, WithViewState } from '@aemforms/af-react-renderer';

export type FieldViewState = WithViewState<FieldJson>;

export type PROPS = State<FieldJson & Handlers & {
  isError?: boolean,
  layout?: {
    [key: string]: any
  }
  text?:string,
  richText?:boolean
}>;

export type PROPS_PANEL = State<FieldsetJson> & { handlers: Handlers }
