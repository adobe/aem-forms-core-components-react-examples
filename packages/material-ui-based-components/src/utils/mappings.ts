import Form from '../components/Form';
import Panel from '../components/Panel';
import TextField from '../components/TextField';
import Date from '../components/Date';
import PlainText from '../components/PlainText';
import Title from '../components/Title';
import Image from '../components/Image';
import Rating from '../components/Rating';
import Slider from '../components/Slider';
import TabsVerticalComp from '../components/TabsVerticalComp';
import TabsHorizontalComp from '../components/TabsHorizontalComp';

export default {
  form: Form,
  panel: Panel,
  'text-input': TextField,
  'date-input': Date,
  'text-input': TextField,
  'plain-text': PlainText,
  'custom:title': Title,
  image: Image,
  'text-input': TextField,
  'custom:rating':Rating,
  'custom:slider':Slider
  'text-input': TextField,
  'custom:vertical-tab': TabsVerticalComp,
  'custom:horizontal-tab': TabsHorizontalComp
};