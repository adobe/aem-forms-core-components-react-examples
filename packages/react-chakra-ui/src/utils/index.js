import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
