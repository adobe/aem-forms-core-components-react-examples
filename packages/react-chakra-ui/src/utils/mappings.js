import TextField from "../generic_components/TextFieldComp";
import DateField from "../components/DateFieldComp";
import ButtonComp from "../generic_components/ButtonComp";
import CheckBoxComp from "../generic_components/CheckBoxComp";
import TextAreaComp from "../generic_components/TextAreaComp";
import Panel from "../generic_components/Panel";
import CustomPanel from "../components/CustomPanel";
import CustomText from "../components/CustomText";
import CustomTextButton from "../components/CustomTextButton";
import CustomHeading from "../components/CustomHeading";
import SelectComp from "../generic_components/SelectComp";
import VerticalContainer from "../components/VerticalContainer";
import HorizontalWrap from "../components/HorizontalWrap";
import CheckBoxGroup from "../components/CheckBoxGroup";
import TabsComp from "../components/TabsComp";
import RadioComp from "../generic_components/RadioComp";
import AccordionComp from "../generic_components/AccordionComp";
import LoanButtonComp from "../components/LoanButtonComp"; 
import CustomTextAreaComp from "../components/CustomTextArea";
import Form from "../components/Form";

const myMapping = {
  "text-input": TextField,
  "date-input": DateField,
  "number-input": TextField,
  "multiline-input": TextAreaComp,
  checkbox: CheckBoxComp,
  "checkbox-group": CheckBoxGroup,
  button: ButtonComp,
  panel: Panel,
  "custom:panel": CustomPanel,
  "custom:loanButton": LoanButtonComp,
  "drop-down": SelectComp,
  "radio-group": RadioComp,
  "custom:static/text": CustomText,
  "custom:static/textButton": CustomTextButton,
  "custom:static/heading": CustomHeading,
  "custom:vertical-flex": VerticalContainer,
  "custom:horizontal-flex": HorizontalWrap,
  "custom:vertical-tab": TabsComp,
  "custom:accordion": AccordionComp,
  "custom:accordionItem": CustomTextAreaComp,
  form: Form
};
export default myMapping;
