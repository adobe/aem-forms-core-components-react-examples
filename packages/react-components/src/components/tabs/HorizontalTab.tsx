import React,  {useContext, useCallback, useState} from 'react';
import {FormContext} from '@aemforms/af-react-renderer';
import {getRenderer, useFormIntl} from '@aemforms/af-react-renderer';
import {ContainerJson, State, CUSTOM_PROPS_KEY, TRANSLATION_ID, getOrElse} from '@aemforms/af-core';

const Tabs = (props: State<ContainerJson>) => {
   console.log('tab-props',props);
   const {mappings} = useContext(FormContext);
   const [activetabIndex, setActiveTabIndex] = useState(0);
   const { items, label, id, visible, enabled } = props;
   const {v: visibleItems} =
   items.reduce(({v} : any, item) => {
       const isVisible = item.visible === true;
       return {
           v: isVisible ? v.concat([item]) : v
       };
   }, {v: []});

const localizeTabLabel = useCallback((item: State<ContainerJson>) => {
   let defaultMessage = item?.label?.value;
   const id = getOrElse(item, [CUSTOM_PROPS_KEY, TRANSLATION_ID, 'label.value']);
   if (id) {
       const i18n = useFormIntl();
       defaultMessage = i18n.formatMessage({id, defaultMessage});
   }
   return defaultMessage;
}, [useFormIntl]);

const handleClick = (index: number) => {
   setActiveTabIndex(index);
}

const Comp = getRenderer(visibleItems[activetabIndex], mappings);

return visible ? (
   <div id={id} className="cmp-tabs" aria-labelledby={`${id}-label`} data-placeholder-text="false" data-cmp-visible={visible?'true':'false'} data-cmp-enabled={enabled?'true':'false'}>
        {label?.visible && <label className="cmp-tabs__label" htmlFor={id} id={`${id}-label`}>{label?.value}</label>}
               <ol role="tablist" className="cmp-tabs__tablist" aria-multiselectable="false" >
                  {visibleItems.map((item: State<ContainerJson>, index: number) =>( 
                  <li id={`${id}-item-${item?.id}-tab`} role="tab" key={item?.id} onClick={() => handleClick(index)} className={index===activetabIndex?'cmp-tabs__tab cmp-tabs__tab--active': 'cmp-tabs__tab'} data-cmp-hook-adaptiveformtabs="tab" aria-selected={index===activetabIndex?'true':'false'} tabIndex={index===activetabIndex?0:-1}>{localizeTabLabel(item)}</li>
                   ))}
              </ol>
               <div id={`${id}-item-${visibleItems[activetabIndex].id}-tabpanel`} data-cmp-hook-adaptiveformtabs="tabpanel" className={activetabIndex === visibleItems[activetabIndex].index?'cmp-tabs__tabpanel cmp-tabs__tabpanel--active':'cmp-tabs__tabpanel'} role="tabpanel" tabIndex={0} aria-hidden={activetabIndex===visibleItems[activetabIndex].index?'false':'true'}>
                  <Comp key={`${visibleItems[activetabIndex].id}_${activetabIndex}`} {...visibleItems[activetabIndex]} />
               </div>
   </div>):null;
};

export default Tabs;