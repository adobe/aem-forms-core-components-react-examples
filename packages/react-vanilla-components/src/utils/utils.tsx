export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};