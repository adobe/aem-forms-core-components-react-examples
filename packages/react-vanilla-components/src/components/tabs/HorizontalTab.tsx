// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  *  The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/tabsontop/v1/tabsontop/tabsontop.html
//  ******************************************************************************

import React, { useContext, useState, useCallback } from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { getRenderer } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../../utils/withRuleEngine';
import { PROPS_PANEL } from '../../utils/type';

const HorizontalTab = (props: PROPS_PANEL) => {
  console.log('tab-props', props);
  // @ts-ignore
  const { mappings } = useContext(FormContext);
  const [activetabIndex, setActiveTabIndex] = useState(0);
  const [shortDescription, setShortDescription] = useState(true);
  const [longDescription, setLongtDescription] = useState(false);
  const { items, label, id, visible, enabled } = props;
  const { v: visibleItems } =
    items.reduce(({ v }: any, item) => {
      const isVisible = item.visible === true;
      return {
        v: isVisible ? v.concat([item]) : v
      };
    }, { v: [] });

  const handleClick = (index: number) => {
    setActiveTabIndex(index);
  }

  const handleButtonClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event.preventDefault();
    setShortDescription(!shortDescription);
    setLongtDescription(!longDescription);
  };

  const getChild = useCallback((child: any, index: number) => {
    const Comp = getRenderer(child, mappings);
    return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
  }, []);

  return (
    <div data-cmp-is="adaptiveFormTabs" id={id} className="cmp-tabs" data-placeholder-text="false" data-cmp-visible={visible} data-cmp-enabled={enabled}>
      {label?.visible && <label className="cmp-tabs__label" htmlFor={`${id}-widget`}>{label?.value}</label>}
      {props?.description && <button aria-label='Toggle Button' className="cmp-tabs__questionmark" onClick={handleButtonClick}></button>}
      {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-tabs__shortdescription'>{props?.tooltip}</div>}
      <div aria-live="polite">
        {longDescription && props?.description && <div title='Help Text' data-cmp-visible={longDescription} className="cmp-tabs__longdescription">{props?.description}</div>}
      </div>
      {visibleItems.length > 0 && <ol role="tablist" className="cmp-tabs__tablist" aria-multiselectable="false" id={`${id}-widget`} >
        {visibleItems.map((item: PROPS_PANEL, index: number) => (
          <li id={`${item?.id}__tab`} aria-controls={`${item?.id}__tabpanel`} role="tab" key={item?.id} onClick={() => handleClick(index)} className={index === activetabIndex ? 'cmp-tabs__tab cmp-tabs__tab--active' : 'cmp-tabs__tab'} data-cmp-hook-adaptiveformtabs="tab" aria-selected={index === activetabIndex ? 'true' : 'false'} tabIndex={index === activetabIndex ? 0 : -1}>{item.label?.value}</li>
        ))}
      </ol>}
      {visibleItems.length > 0 ? <div id={`${visibleItems[activetabIndex].id}__tabpanel`} data-cmp-hook-adaptiveformtabs="tabpanel" className={activetabIndex === visibleItems[activetabIndex].index ? 'cmp-tabs__tabpanel cmp-tabs__tabpanel--active' : 'cmp-tabs__tabpanel'} role="tabpanel" tabIndex={0} aria-hidden={activetabIndex === visibleItems[activetabIndex].index ? 'false' : 'true'} aria-labelledby={`${visibleItems[activetabIndex].id}__tab`}>
        {getChild(visibleItems[activetabIndex], activetabIndex)}
      </div> : null}
    </div>);
};

export default withRuleEnginePanel(HorizontalTab);