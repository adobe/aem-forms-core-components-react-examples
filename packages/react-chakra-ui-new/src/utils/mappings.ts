// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import Button from '../components/Button';
import CheckBox from '../components/CheckBox';
import CheckBoxGroup from '../components/CheckBoxGroup';
import DropDown from '../components/DropDown';
import Number from '../components/Number';
import RadioGroup from '../components/RadioButton';
import TextField from '../components/TextField';
import Panel from '../components/Panel';
import Wizard from '../components/Wizard';
import Accordion from '../components/Accordion';
import HorizontalTab from '../components/HorizontalTabs';
import VerticalTab from '../components/VerticalTabs';
import Form from '../components/Form';
import Image from '../components/Image';
import PlainText from '../components/PlainText';

const mappings = {
  'text-input': TextField,
  'password-input': TextField,
  'multiline-input': TextField,
  panel: Panel,
  'checkbox-group': CheckBoxGroup,
  'radio-group': RadioGroup,
  'drop-down': DropDown,
  'number-input': Number,
  button: Button,
  checkbox: CheckBox,
  image: Image,
  'core/fd/components/form/wizard/v1/wizard': Wizard,
  'core/fd/components/form/accordion/v1/accordion': Accordion,
  'core/fd/components/form/tabsontop/v1/tabsontop': HorizontalTab,
  'core/fd/components/form/verticaltabs/v1/verticaltabs': VerticalTab,
  'plain-text': PlainText,
  'core/fd/components/form/panelcontainer/v1/panelcontainer': Panel,
  form: Form
};

export default mappings;