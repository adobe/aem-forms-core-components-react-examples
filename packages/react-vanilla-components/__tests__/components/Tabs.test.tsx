/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import HorizontalTab from "../../src/components/tabs/HorizontalTab";
import React from "react";
import { render } from "@testing-library/react";
import { renderComponent } from "../utils";
import userEvent from "@testing-library/user-event";

const emptyTabs = {
  id: "emptyTabs",
  ":type": "core/fd/components/form/tabsontop/v1/tabsontop",
  visible: true,
  label: { value: "empty label" },
  items: [],
  index: 0,
};

const tabsWithData = {
  id: "tabsWithData",
  ":type": "core/fd/components/form/tabsontop/v1/tabsontop",
  visible: true,
  label: { value: "tab label" },
  index: 0,
  items: [
    {
      id: "field",
      name: "f1",
      fieldType: "text-field",
      visible: true,
      index: 0,
      label: { value: "text field label" },
    },
    {
      id: "field2",
      name: "f2",
      fieldType: "text-field",
      visible: true,
      index: 1,
      label: { value: "text field 2 label" },
    },
  ],
};

const helper = renderComponent(HorizontalTab);

test("Tabs should not rendered if item length is zero", () => {
  const { renderResponse } = helper(emptyTabs);
  expect(renderResponse.queryByText(emptyTabs.label.value)).not.toBeNull();
  const tabs = renderResponse.queryAllByRole("tablist");
  expect(tabs.length).toEqual(0);
  const tabPanels = renderResponse.queryAllByRole("tabpanel");
  expect(tabPanels.length).toEqual(0);
});

test("Tabs should not rendered if visible is false", () => {
  const sample = {
    ...emptyTabs,
    visible: false,
  };
  const { container } = render(<HorizontalTab {...sample} />);
  expect(container.innerHTML.length).toEqual(0);
});

test("clicking on the horizontal tabs should switch from one tab to another",() => {
  const { renderResponse } = helper(tabsWithData);
  const tablist = renderResponse.getAllByRole('tab');
  expect(tablist).toHaveLength(2);
  expect(tablist[0].className.includes('cmp-tabs__tab cmp-tabs__tab--active'))
  expect(tablist[1].className.includes('cmp-tabs__tab'))
  userEvent.click(tablist[1]);
  expect(tablist[0].className.includes('cmp-tabs__tab'))
  expect(tablist[1].className.includes('cmp-tabs__tab cmp-tabs__tab--active'))
});

test("tabs should get rendered with undefined element if no mapping is defined", () => {
  const { renderResponse } = helper(tabsWithData);
  const tabs = renderResponse.container.querySelectorAll('[role="tab"]');
  expect(tabs.length).toEqual(2);
  const tabPanels = renderResponse.container.querySelectorAll(".undefined-element");
  expect(tabPanels.length).toEqual(0);
});

test("hidden tabs should not be rendered", () => {
  const sample = {
    ...tabsWithData,
    items: [
      {
        ...tabsWithData.items[0],
        visible: false,
      },
      {
        ...tabsWithData.items[1],
      },
    ],
  };
  const { renderResponse } = helper(sample);
  const tabs = renderResponse.container.querySelectorAll('[role="tab"]');
  expect(tabs.length).toEqual(1);
});

test("Tabs should not rendered if all the items are not visible", () => {
  const sample = {
    ...tabsWithData,
    items: tabsWithData.items.map((x) => ({
      ...x,
      visible: false,
    })),
  };
  const { renderResponse } = helper(sample);
  expect(renderResponse.queryByText(sample.label.value)).not.toBeNull();
  const tabs = renderResponse.queryAllByRole("tab");
  expect(tabs.length).toEqual(0);
  const tabPanels = renderResponse.queryAllByRole("tabpanel");
  expect(tabPanels.length).toEqual(0);
});

test('html in the label should be handled for non rich text', () => {
  const f = {
    ...tabsWithData,
    label: {
      value: '<p>title inside p tags</p>',
      richText: true,
      visible: true
    }
  }
  let { renderResponse } = helper(f);
  expect(renderResponse.container.innerHTML).toContain('<p>title inside p tags</p>');
});
