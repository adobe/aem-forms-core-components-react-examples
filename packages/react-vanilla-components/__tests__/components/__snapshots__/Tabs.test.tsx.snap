// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <div
    class="cmp-tabs "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormTabs"
    data-cmp-visible="true"
    data-placeholder-text="false"
    id="tabsWithData"
  >
    <div
      class="cmp-tabs__label-container"
    >
      <label
        class="cmp-tabs__label"
        for="tabsWithData-widget"
      >
        tab label
      </label>
    </div>
    <div
      aria-live="polite"
    />
    <ol
      aria-multiselectable="false"
      class="cmp-tabs__tablist"
      id="tabsWithData-widget"
      role="tablist"
    >
      <li
        aria-controls="field__tabpanel"
        aria-selected="true"
        class="cmp-tabs__tab cmp-tabs__tab--active"
        data-cmp-hook-adaptiveformtabs="tab"
        id="field__tab"
        role="tab"
        tabindex="0"
      >
        text field label
      </li>
      <li
        aria-controls="field2__tabpanel"
        aria-selected="false"
        class="cmp-tabs__tab"
        data-cmp-hook-adaptiveformtabs="tab"
        id="field2__tab"
        role="tab"
        tabindex="-1"
      >
        text field 2 label
      </li>
    </ol>
    <div
      aria-hidden="false"
      aria-labelledby="field__tab"
      class="cmp-tabs__tabpanel cmp-tabs__tabpanel--active"
      data-cmp-hook-adaptiveformtabs="tabpanel"
      id="field__tabpanel"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
